test/csuite/config/main.c

```diff
@@ -0,0 +1,494 @@
+        dest = _dest;                          \
+    } while (0)
+
+/*
+ * For various standard types, a list of valid and invalid strings to try.
+ */
+static const char *valid_bool[] = {"true", "false", "0", "1", "", NULL};
+static const char *invalid_bool[] = {"foo", "TRUE", "FALSE", "-2", NULL};
+
+static const char *valid_string[] = {"foo", "", "42", "-42", NULL};
+static const char *invalid_string[] = {"+$", NULL};
+
+static const char *valid_nonnegative_int[] = {"42", "0", NULL};
+static const char *invalid_nonnegative_int[] = {"foo", "", "-42", NULL};
+
+/*
+ * Check compiling configuration strings for begin_transaction.
+ *
+ *   "ignore_prepare=%s,isolation=%s,name=%s,no_timestamp=%s,"
+ *   "operation_timeout_ms=%s,priority=%s,read_timestamp=%s,"
+ *   "roundup_timestamps=(prepared=%s,read=%s),sync=%s";
+ */
+
+/*
+ * These are special valid/invalid pairs used with specific keys in configuration for
+ * begin_transaction. Every choice list implicitly allows a blank value.
+ */
+static const char *valid_ignore_prepare[] = {"true", "false", "force", "", NULL};
+static const char *invalid_ignore_prepare[] = {"foo", NULL};
+
+static const char *valid_isolation[] = {"read-uncommitted", "read-committed", "snapshot", "", NULL};
+static const char *invalid_isolation[] = {"foo", NULL};
+
+/* -100 to 100 allowed */
+static const char *valid_priority_int[] = {"0", "-100", "100", NULL};
+static const char *invalid_priority_int[] = {"foo", "-101", "101", NULL};
+
+/*
+ * The complete set of configuration keys for begin_transaction, along with valid/invalid pairs for
+ * each.
+ */
+static KEY_VALUES begin_txn_roundup_kv[] = {{"prepared", valid_bool, invalid_bool, NULL},
+  {"read", valid_bool, invalid_bool, NULL}, {NULL, NULL, NULL, NULL}};
+static KEY_VALUES begin_txn_kv[] = {
+  {"ignore_prepare", valid_ignore_prepare, invalid_ignore_prepare, NULL},
+  {"isolation", valid_isolation, invalid_isolation, NULL},
+  {"name", valid_string, invalid_string, NULL}, {"no_timestamp", valid_bool, invalid_bool, NULL},
+  {"operation_timeout_ms", valid_nonnegative_int, invalid_nonnegative_int, NULL},
+  {"priority", valid_priority_int, invalid_priority_int, NULL},
+  /*
+   * As far as the parser is concerned, the read timestamp is just a string. It is validated by each
+   * API configuration code. This is likely to change, we'll probably introduce a timestamp type,
+   * until then, this is the best we can do.
+   */
+  {"read_timestamp", valid_string, invalid_string, NULL},
+  {"roundup_timestamps", NULL, NULL, begin_txn_roundup_kv},
+  {"sync", valid_bool, invalid_bool, NULL}, {NULL, NULL, NULL, NULL}};
+
+/*
+ * free_parse_state --
+ *     Free and zero the state to prepare for another set of callbacks.
+ */
+static void
+free_parse_state(PARSE_STATE *state)
+{
+    int i;
+
+    free(state->method_name);
+    state->method_name = NULL;
+    for (i = 0; i < MAX_INPUT_CONFIG; ++i) {
+        free(state->input_config[i]);
```

Comment:
Are we sure that input config is always not NULL?

Reply:
`input_config` entries may be NULL, but `free(NULL)` must be accepted by a conforming C compiler.